tincr/io/device/family_info.tcl

```diff
@@ -0,0 +1,773 @@
+package provide tincr.io.device 0.0
+package require Tcl 8.5
+package require tincr.cad.design 0.0
+package require tincr.cad.device 0.0
+package require tincr.cad.util 0.0
+
+namespace eval ::tincr:: {
+    namespace export \
+	create_xml_family_info 
+}
+
+# #########################################
+#           Family Info Notes:
+# ######################################### 
+# This Tcl script is used to generate FamilyInfo.xml files for RapidSmith2 or other external tools.
+# The Family Info augments the XDLRC file with additional information about devices. Currently,
+# some hand-edits are required for things that cannot be automatically generated using Vivado Tcl
+# commands.
+#
+# Series7:
+# -------
+# 1.) Alternate site pin mappings: When a site type is changed in Vivado to an alternate type, the
+#     pins may be renamed. We cannot determine the which pins are renamed for series7 devices, and
+#     so the user must do this manually. NOT APPICABLE TO ULTRASCALE
+# 2.) Some sites in series 7 return invalid alternate types. Invalid alternate types need to be removed by hand
+#     by looking at the alternate types in the device browser GUI
+# 3.) The series7 primitive defs in Tincr contain sites that do not appear in the Tcl interface. At some point, we need
+#     to remove these primitive defs, but for now we can add them to the end of the family info with limited info.
+# 4.) Compatible types need to be added for compatible sites with more than one BEL (generally SLICEL->SLICEM  and
+#     IOB sites)
+#
+# UltraScale:
+# -----------
+# 1.) Compatible types need to be added for compatible sites with more than one BEL (generally SLICEL->SLICEM  and
+#     IOB sites). Most compatible types for single-BEL sites can be generated though.
+#     - Under SLICEL
+#     <compatible_types>
+#       <compatible_type>SLICEM</compatible_type>
+#     </compatible_types>
+#
+#     - UNDER HRIO
+#     <compatible_types>
+#       <compatible_type>HPIOB</compatible_type>
+#     </compatible_types>
+
+# global variable set to 1 if a family is within the series7 of devices
+# Global variables
+set is_series7 0
```

Comment:
Best to avoid globals if possible. How much time are you really saving by computing this just once as opposed to having each proc compute it?